File: clients/js/src/features/config_resolver.ts
import { Input } from "../input";

import {
  ClientConfigBuilder,
  PolywrapClient,
  IUriResolver,
  Uri
} from "@polywrap/client-js";

export async function runTestCase(input: unknown): Promise<void> {
  const inputObj = Input.expectObject<{
    authority: unknown;
    result: unknown;
  }>(input);
  const authority = Input.expectString(inputObj.authority);
  const result = Input.expectString(inputObj.result);

  console.log("Adding Resolver to ClientConfig");

  const resolver: IUriResolver<string> = {
    tryResolveUri: async (uri: Uri) => {
      const response = uri.authority === authority
        ? Uri.from(result)
        : uri;

      return {
        ok: true,
        value: {
          type: "uri",
          uri: response
        }
      };
    }
  }

  const config = new ClientConfigBuilder()
    .addResolver(resolver)
    .build();

  const client = new PolywrapClient(config);

  console.log(`Resolving a wrap://${authority} URI`);

  const res = await client.tryResolveUri({
    uri: Uri.from(`wrap://${authority}/foo`)
  });

  console.log(res)

  if (res.ok && res.value.type === "uri") {
    console.log(`Received URI '${res.value.uri}'`);
    console.log("Success!");
  }
}

File: clients/py/src/features/config_resolver.py
from typing import Any

from polywrap import (
    PolywrapClient,
    PolywrapClientConfigBuilder,
    Uri,
    UriPackageOrWrapper,
    UriResolver,
)
from pydantic import BaseModel

from validators import UriStr


class TestCaseInput(BaseModel):
    authority: str
    result: UriStr


class CustomResolver(UriResolver):
    def __init__(self, authority: str, result: str):
        self.authority = authority
        self.result = result

    def try_resolve_uri(self, uri: Uri, *_: Any) -> UriPackageOrWrapper:
        return Uri.from_str(self.result) if uri.authority == self.authority else uri


def run_test_case(input: Any) -> None:
    input_obj = TestCaseInput.parse_obj(input)

    print("Adding Resolver to ClientConfig")

    resolver = CustomResolver(input_obj.authority, input_obj.result.uri)

    config = PolywrapClientConfigBuilder().add_resolver(resolver).build()

    client = PolywrapClient(config)

    print(f"Resolving a wrap://{input_obj.authority} URI")

    res = client.try_resolve_uri(Uri.from_str(f"wrap://{input_obj.authority}/foo"))

    match res:
        case Uri() as uri:
            print(f"Received URI '{uri}'")
            print("Success!")
        case _:
            print("Failed to resolve URI.")
